### PR TITLE
Complete day 2 challenge

### PR DESCRIPTION
**Pull request by mrseanbaines**

### CODE REVIEW

File: 2020/day-2-password-philosophy/index.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  return list
+    .trim()
+    .split('\n')
+    .map(entry => entry.split(': '))
+    .map(([policy, password]) => {
+      const [range, char] = policy.trim().split(' ')
+      const [min, max] = range.split('-').map(Number)
+
+      return { min, max, char, password }
+    })
```
Comment (simonfl3tcher) [inline 2020/day-2-password-philosophy/index.ts:11]: This is way more idiomatic than my implementation. It's really easy to read as well, nice work.

File: 2020/day-2-password-philosophy/index.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+export const getValidPasswords1 = (list: string) => {
+  const entries = parseList(list)
+
+  return entries.filter(({ min, max, char, password }) => {
+    const charCount = Array.from(password.matchAll(new RegExp(char, 'g'))).length
```
Comment (simonfl3tcher) [inline 2020/day-2-password-philosophy/index.ts:18]: Oh this is really neat!

File: 2020/day-2-password-philosophy/index.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+export const getValidPasswords2 = (list: string) => {
+  const entries = parseList(list)
+
+  return entries.filter(({ min, max, char, password }) => {
+    const firstChar = password.charAt(min - 1)
+    const lastChar = password.charAt(max - 1)
+
+    return [firstChar, lastChar].filter(c => c === char).length === 1
+  })
```
Comment (simonfl3tcher) [inline 2020/day-2-password-philosophy/index.ts:32]: This is really elegant baines, I love the way you have thought about this.

File: 2020/day-2-password-philosophy/index.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    .map(([policy, password]) => {
+      const [range, char] = policy.trim().split(' ')
+      const [min, max] = range.split('-').map(Number)
+
+      return { min, max, char, password }
+    })
+}
+
+export const getValidPasswords1 = (list: string) => {
+  const entries = parseList(list)
```
Comment (simonfl3tcher) [inline 2020/day-2-password-philosophy/index.ts:15]: I wonder if it makes sense to put this in a const outside of the methods so we don't need to duplicate it?

File: 2020/day-2-password-philosophy/index.ts (hunk 2 above)
Comment (mrseanbaines) [reply 2020/day-2-password-philosophy/index.ts:18]: I think `.matchAll` returned an iterator or something, so `Array.from` converts it into a standard array. Could also use the spread: `[...password.matchAll(new RegExp(char, 'g'))].length`

File: 2020/day-2-password-philosophy/index.ts (hunk 4 above)
Comment (mrseanbaines) [reply 2020/day-2-password-philosophy/index.ts:15]: Hm, but we need to pass `list` to `parseList`, which is currently only available as a parameter of these `getValidPasswords*` functions

**Pull request merged**